### PR TITLE
Add flatten op

### LINKED ISSUE #12340: Need add flatten op
Need add flatten op, to do reshape in detection api

### PR DESCRIPTION
**Pull request by baiyfbupt**
resolve #12340

### CODE REVIEW

File: paddle/fluid/operators/flatten_op.cc [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+                 "the input tensor. When axis = 0, the shape of the output"
+                 "tensor is (1, (d_0 X d_1 ... d_n), where the shape of the"
+                 "input tensor is (d_0, d_1, ... d_n).")
+        .SetDefault(1);
+    AddAttr<bool>("inplace",
+                  "(default: false) flatten the source tensor without "
+                  "memory copy. When Attr(inplace) is set true, the output "
+                  "tensor shares memory with Input(X), otherwise, a new output "
+                  "tensor is created, and its data are copied from Input(x).")
+        .SetDefault(false);
```
Comment (qingqing01) [inline paddle/fluid/operators/flatten_op.cc]: Remove this attr. 这个在face detection实验中，加内存优化有bug。 就不用留给用户了。

File: paddle/fluid/operators/flatten_op.cc [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+
+class FlattenOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  void Make() override {
+    AddInput("X", "(Tensor) A tensor of rank >= axis.");
+    AddOutput("Out",
+              "A 2D tensor with the contents of the input tensor,"
+              "with input dimensions up to axis flattened to the outer"
+              "dimension of the output and remaining input dimensions"
+              "flattened into the inner dimension of the output.");
```
Comment (qingqing01) [inline paddle/fluid/operators/flatten_op.cc]: 语法不对。

A 2D tensor is reshaped input tensor. The input dimensions up to axis are flattened to the outer dimension of the output and the remaining input dimensions are flattened into the inner dimension of the output.

File: paddle/fluid/operators/flatten_op.cc [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+ private:
+  void RunImpl(const framework::Scope &scope,
+               const platform::Place &place) const override {
+    auto dx_name = Output(framework::GradVarName("X"));
+    auto dout_name = Input(framework::GradVarName("Out"));
+    auto in_dims =
+        scope.FindVar(Input("X"))->Get<framework::LoDTensor>().dims();
+    framework::AttributeMap attrs;
+    attrs["shape"] = framework::vectorize2int(in_dims);
+    attrs["inplace"] = Attr<bool>("inplace");
```
Comment (qingqing01) [inline paddle/fluid/operators/flatten_op.cc]:
```c++
attrs["inplace"] = false;
```

File: paddle/fluid/operators/flatten_op.cc [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+  void RunImpl(const framework::Scope &scope,
+               const platform::Place &place) const override {
+    auto &axis = Attr<int>("axis");
+    auto in_dims =
+        scope.FindVar(Input("X"))->Get<framework::LoDTensor>().dims();
+    auto out_dims = FlattenOpInferShape::GetOutputShape(axis, in_dims);
+
+    framework::AttributeMap attrs;
+    attrs["shape"] = framework::vectorize2int(out_dims);
+    attrs["inplace"] = Attr<bool>("inplace");
```
Comment (qingqing01) [inline paddle/fluid/operators/flatten_op.cc]:
```
attrs["inplace"] = false;
```

File: paddle/fluid/operators/flatten_op.cc [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+ private:
+  void RunImpl(const framework::Scope &scope,
+               const platform::Place &place) const override {
+    auto &axis = Attr<int>("axis");
+    auto in_dims =
+        scope.FindVar(Input("X"))->Get<framework::LoDTensor>().dims();
+    auto out_dims = FlattenOpInferShape::GetOutputShape(axis, in_dims);
+
+    framework::AttributeMap attrs;
+    attrs["shape"] = framework::vectorize2int(out_dims);
```
Comment (qingqing01) [inline paddle/fluid/operators/flatten_op.cc]: `GetOutputShape`直接返回`std::vector<int64_t>`即可， 这里也不用转换了吧。

File: paddle/fluid/operators/flatten_op.cc (hunk 1 above)
Comment (baiyfbupt) [reply paddle/fluid/operators/flatten_op.cc]: Done,thanks!

File: paddle/fluid/operators/flatten_op.cc (hunk 2 above)
Comment (baiyfbupt) [reply paddle/fluid/operators/flatten_op.cc]: Done,thanks!

File: paddle/fluid/operators/flatten_op.cc (hunk 3 above)
Comment (baiyfbupt) [reply paddle/fluid/operators/flatten_op.cc]: Done,thanks!

File: paddle/fluid/operators/flatten_op.cc (hunk 4 above)
Comment (baiyfbupt) [reply paddle/fluid/operators/flatten_op.cc]: Done,thanks!

File: paddle/fluid/operators/flatten_op.cc (hunk 5 above)
Comment (baiyfbupt) [reply paddle/fluid/operators/flatten_op.cc]: Done,thanks!

**Review by qingqing01 — APPROVED** [review]

**Pull request merged**